state-chain/pallets/cf-broadcast/src/tests.rs

```diff
@@ -625,3 +625,72 @@ fn retry_and_success_in_same_block() {
 			},
 		);
 }
+
+// When we retry threshold signing, we want to make sure that the storage remains valid such that if
+// there is transaction_succeeded witnessed late due to some delay, the success still goes through.
+// We use the second attempt to ensure that we are not pulling the default of 0 for `ValueQuery` of
+// `BroadcastAttemptCount`.
+// Note: At time of writing there is a bug here, that the tx fee is not refunded in the case we
+// re-threshold sign and then witness success.
+#[test]
+fn retry_with_threshold_signing_still_allows_late_success_witness_second_attempt() {
+	new_test_ext().execute_with(|| {
+		let broadcast_attempt_id = start_mock_broadcast_tx_out_id(MOCK_TRANSACTION_OUT_ID);
+
+		let awaiting_broadcast =
+			AwaitingBroadcast::<Test, Instance1>::get(broadcast_attempt_id).unwrap();
+
+		assert_eq!(
+			BroadcastAttemptCount::<Test, Instance1>::get(broadcast_attempt_id.broadcast_id),
+			0
+		);
+		let nominee = MockNominator::get_last_nominee().unwrap();
+
+		let current_block = frame_system::Pallet::<Test>::block_number();
+
+		let expected_expiry_block = current_block + BROADCAST_EXPIRY_BLOCKS;
+
+		assert_eq!(
+			Timeouts::<Test, Instance1>::get(expected_expiry_block)
+				.into_iter()
+				.next()
+				.unwrap(),
+			broadcast_attempt_id
+		);
+
+		// We want to run test test on the second attempt.
+		MockCfe::respond(Scenario::SigningFailure);
+		Broadcaster::on_idle(0, LARGE_EXCESS_WEIGHT);
+		assert_eq!(
+			BroadcastAttemptCount::<Test, Instance1>::get(broadcast_attempt_id.broadcast_id),
+			1
+		);
+
+		MockTransactionBuilder::<MockEthereum, RuntimeCall>::set_invalid_for_rebroadcast();
+
+		// Retry will be triggered on timeout, and it'll take the invalid signature code path.
+		Broadcaster::on_initialize(expected_expiry_block);
```

Comment:
If possible it would be better to use the `execute_at_block` test util since it also runs the on_idle hook - which is what triggers retries. In this test we run on_idle before on_initialize, which only reflects the real order of execution if we assume these things happen on subsequent blocks.